### PR TITLE
Federated Analytics and Histogram Taskrunner Workspace

### PR DESCRIPTION
**Pull request by tanwarsh**
## Summary
The PR contains changes for Federated Analytics and Histogram Taskrunner Workspace
as discussed in https://github.com/securefederatedai/openfl/discussions/1385#discussioncomment-12548664

## Type of Change (Mandatory)
Specify the type of change being made. 
- New Feature.  

## Description (Mandatory)
The PR includes following changes:

- Core component level changes with mode analysis for FA, enabling users to run federated analytics with the OpenFL Taskrunner API.
- Collaborators will query data and return analysis results.
- The aggregator will aggregate the analysis results.
- The Base Taskrunner class will be used for FA.
- Taskrunner Workspace: Calculate a histogram (an estimate of the probability distribution of a continuous variable) on specific columns of the [Iris](https://huggingface.co/datasets/scikit-learn/iris)[ dat](https://huggingface.co/datasets/scikit-learn/iris)[aset](https://huggingface.co/datasets/scikit-learn/iris).

## Testing
Screenshot 2025-03-21 at 6 32 16 PM
Screenshot 2025-03-21 at 6 32 23 PM

## Additional Information
1. Changed workspace form torch/mnist to keras/mnist for secure aggregation pipeline due to issue https://github.com/securefederatedai/openfl/issues/1475.
2. Ignore error - openfl/interface/plan.py:100:5: C901 `initialize` is too complex (11 > 10)

### CODE REVIEW

**Comment by tanwarsh** [comment]
Working on resolving issues with pipelines:
1. [TaskRunner Secure Aggregation E2E / Secure Aggregation (torch/mnist, 3.10)](https://github.com/securefederatedai/openfl/actions/runs/13992442997/job/39240952448?pr=1465#logs)
2. [Pytest and code coverage / build](https://github.com/securefederatedai/openfl/actions/runs/14027269853/job/39268046688?pr=1465#logs)

File: openfl/component/aggregator/aggregator.py [hunk 1]
```diff
@@ -86,6 +87,7 @@ def __init__(
         persist_checkpoint=True,
         persistent_db_path=None,
         secure_aggregation=False,
+        mode="learning",
```
Comment (payalcha) [inline openfl/component/aggregator/aggregator.py:90]: If we are introducing mode as separate field, are we planning to change "evaluation" as well. Else it will be confusing.

File: openfl/component/aggregator/aggregator.py [hunk 2]
```diff
@@ -132,18 +134,59 @@ def __init__(
         if self.assigner.is_task_group_evaluation():
```
Comment (payalcha) [inline openfl/component/aggregator/aggregator.py:134]: Similar change we are doing in else condition of mode if not learning

**Review by teoparvanov — CHANGES_REQUESTED** [review]
Hi @tanwarsh, there seems to be some disconnect here... My understanding from the design discussion was that for phase 1, you would simply introduce specialized task runner/data loader for FA, serving a specific query, hard-coded in the FL plan. Why do the core framework's collaborator and aggregator need to be aware that they're doing "analytics" (specifically, I don't understand the need for the `mode` parameter at this stage)?

Ideally, in phase 1, you would simply introduce the following artefacts:
- FA task runner
- FA data loader
- Histogram aggregation function (unless it's possible to reuse/extend `FedAvg`?)
- Example FA workspace (incl. FL plan with a "hard-coded" query)

I'd imagine little to no changes required in the core OpenFL framework for the above to work.

On the other hand, natively integrating FA in the core OpenFL framework should be out of scope at this stage. 
Please correct me if I'm missing something.

File: .github/workflows/task_runner_secure_agg_e2e.yml [hunk 3]
```diff
@@ -12,15 +12,15 @@ permissions:
 
 jobs:
   test_secure_aggregation:
-    name: Secure Aggregation (torch/mnist, 3.10)
+    name: Secure Aggregation (keras/mnist, 3.10)
```
Comment (noopurintel) [inline .github/workflows/task_runner_secure_agg_e2e.yml]: Let's keep the model same until unless discussed otherwise.

File: openfl-workspace/federated_analytics/histogram/plan/cols.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (noopurintel) [inline openfl-workspace/federated_analytics/histogram/plan/cols.yaml]: Generic comment: `2020-2025` in all the newly added files.

File: openfl-workspace/federated_analytics/histogram/plan/data.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Copyright (C) 2020-2021 Intel Corporation
+# Licensed subject to the terms of the separately executed evaluation license agreement between Intel Corporation and you.
+
+# collaborator_name,data_directory_path
+one,1
+
+
```
Comment (noopurintel) [inline openfl-workspace/federated_analytics/histogram/plan/data.yaml]: 1 line break is enough.

File: openfl-workspace/federated_analytics/histogram/plan/cols.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+# Licensed subject to the terms of the separately executed evaluation license agreement between Intel Corporation and you.
+
+collaborators:
+
```
Comment (noopurintel) [inline openfl-workspace/federated_analytics/histogram/plan/cols.yaml]: Please remove tab spacing, if present. Due to copy paste between the model workspaces, most of the `cols.yaml` files have tab spacing today, which is unnecessary.

File: openfl-workspace/federated_analytics/histogram/src/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        Loads data shards from a CSV file.
+        This method reads the data from a CSV file located at './data/client.csv'
+        and returns it as a pandas DataFrame.
+        Returns:
+            pd.DataFrame: The data loaded from the CSV file.
+        """
+
+        return pd.read_csv('./data/client.csv')
+
```
Comment (noopurintel) [inline openfl-workspace/federated_analytics/histogram/src/dataloader.py:48]: Single line break is fine.

File: openfl-workspace/federated_analytics/histogram/src/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        Returns:
+            pandas.DataFrame: The shard of the MNIST dataset corresponding to the given shard number.
+        """
+
+        return self._load_datashards().iloc[shard_num::collaborator_count]
+
+    def query(self, columns, **kwargs):
+        """
+        Query the data shard for the specified columns.
+        Parameters:
```
Comment (noopurintel) [inline openfl-workspace/federated_analytics/histogram/src/dataloader.py]: Parameters -> Args or vice versa but let's keep same name in both the functions `load_mnist_shard` and `query`.

File: openfl/component/aggregator/aggregator.py (hunk 1 above)
Comment (tanwarsh) [reply openfl/component/aggregator/aggregator.py:90]: This is not planned as of now. This can be discussed later once these changes are approved.

File: openfl/component/aggregator/aggregator.py (hunk 2 above)
Comment (tanwarsh) [reply openfl/component/aggregator/aggregator.py:134]: Noted.

File: .github/workflows/task_runner_secure_agg_e2e.yml (hunk 3 above)
Comment (tanwarsh) [reply .github/workflows/task_runner_secure_agg_e2e.yml]: this was done because of https://github.com/securefederatedai/openfl/issues/1475.

File: openfl-workspace/federated_analytics/histogram/src/dataloader.py (hunk 7 above)
Comment (tanwarsh) [reply openfl-workspace/federated_analytics/histogram/src/dataloader.py:48]: resolved in [8f95930](https://github.com/securefederatedai/openfl/pull/1465/commits/8f95930abe8337553d167207fbd5250160247f3f).

File: openfl-workspace/federated_analytics/histogram/src/dataloader.py (hunk 8 above)
Comment (tanwarsh) [reply openfl-workspace/federated_analytics/histogram/src/dataloader.py]: resolved in [8f95930](https://github.com/securefederatedai/openfl/pull/1465/commits/8f95930abe8337553d167207fbd5250160247f3f).

File: openfl-workspace/federated_analytics/histogram/plan/cols.yaml (hunk 6 above)
Comment (tanwarsh) [reply openfl-workspace/federated_analytics/histogram/plan/cols.yaml]: resolved in [8f95930](https://github.com/securefederatedai/openfl/pull/1465/commits/8f95930abe8337553d167207fbd5250160247f3f).

File: openfl-workspace/federated_analytics/histogram/plan/data.yaml (hunk 5 above)
Comment (tanwarsh) [reply openfl-workspace/federated_analytics/histogram/plan/data.yaml]: resolved in [8f95930](https://github.com/securefederatedai/openfl/pull/1465/commits/8f95930abe8337553d167207fbd5250160247f3f).

File: openfl-workspace/federated_analytics/histogram/plan/cols.yaml (hunk 4 above)
Comment (tanwarsh) [reply openfl-workspace/federated_analytics/histogram/plan/cols.yaml]: resolved in [8f95930](https://github.com/securefederatedai/openfl/pull/1465/commits/8f95930abe8337553d167207fbd5250160247f3f).

File: .github/workflows/task_runner_secure_agg_e2e.yml (hunk 3 above)
Comment (tanwarsh) [reply .github/workflows/task_runner_secure_agg_e2e.yml]: reverted.

**Comment by tanwarsh** [comment]
> Hi @tanwarsh, there seems to be some disconnect here... My understanding from the design discussion was that for phase 1, you would simply introduce specialized task runner/data loader for FA, serving a specific query, hard-coded in the FL plan. Why do the core framework's collaborator and aggregator need to be aware that they're doing "analytics" (specifically, I don't understand the need for the `mode` parameter at this stage)?
> 
> Ideally, in phase 1, you would simply introduce the following artefacts:
> 
> * FA task runner
> * FA data loader
> * Histogram aggregation function (unless it's possible to reuse/extend `FedAvg`?)
> * Example FA workspace (incl. FL plan with a "hard-coded" query)
> 
> I'd imagine little to no changes required in the core OpenFL framework for the above to work.
> 
> On the other hand, natively integrating FA in the core OpenFL framework should be out of scope at this stage. Please correct me if I'm missing something.

Hi @teoparvanov, Thanks for the review.
1. Core changes are required for this phase. This was discussed in the end of the last meeting as well as this was summarised in the first point here https://github.com/securefederatedai/openfl/discussions/1385#discussioncomment-12548664. Let me know if you still have concerns regarding this.
2. This phase contains changes required to run this basic workspace for FA. this includes taskrunner and loader changes.
3. Histogram aggregation function is required and we cannot use `FedAvg`.
4. I will address the query in plan in future commit.

File: openfl/databases/tensor_db.py [hunk 9]
```diff
@@ -248,15 +249,23 @@ def get_aggregated_tensor(
...
+                for col_name in collaborator_names
+            ]
+        else:
+            local_tensors = [
+                FALocalTensor(
+                    col_name=col_name,
+                    tensor=agg_tensor_dict[col_name],
+                )
+                for col_name in collaborator_names
+            ]
```
Comment (MasterSkepticista) [inline openfl/databases/tensor_db.py:268]: This does not belong in TensorDB. This class is meant for CRUD operations only. We plan on also removing the current `local_tensors` loop as it does not belong here. Please try if this be taken care outside this module

File: openfl/federated/plan/plan.py [hunk 10]
```diff
@@ -477,8 +481,9 @@ def get_task_runner(self, data_loader):
         if self.runner_ is None:
             self.runner_ = Plan.build(**defaults)
 
-        # Define task dependencies after taskrunner has been initialized
-        self.runner_.initialize_tensorkeys_for_functions()
+        if self.config["aggregator"][SETTINGS].get("mode", "learning") == "learning":
+            # Define task dependencies after taskrunner has been initialized
+            self.runner_.initialize_tensorkeys_for_functions()
```
Comment (MasterSkepticista) [inline openfl/federated/plan/plan.py:486]: Can this be taken care within the initialize function? Please try not adding if-else unless absolutely required. This increases the (already high) cyclomatic complexity of the entire codebase.

File: openfl/federated/task/runner_fa.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+"""
+Base classes for Federated Analytics.
+
+You may copy this file as the starting point of your own keras model.
+"""
+
+from openfl.federated.task.runner import TaskRunner
+
+
+class FederatedAnalyticsTaskRunner(TaskRunner):
```
Comment (MasterSkepticista) [inline openfl/federated/task/runner_fa.py:14]: Given the inheritance, I don't think we need to inherit all methods as well, especially when all are passthroughs. It is implied...

File: openfl/interface/plan.py [hunk 12]
```diff
@@ -136,39 +136,40 @@ def initialize(
...
+            init_state_path = plan.config["aggregator"]["settings"]["init_state_path"]
+            # This is needed to bypass data being locally available
+            if input_shape is not None:
+                logger.info(
+                    f"Attempting to generate initial model weights with custom shape {input_shape}"
+                )
+
+            # Initialize tensor dictionary
+            init_tensor_dict, task_runner, round_number = _initialize_tensor_dict(
+                plan, input_shape, init_model_path
```
Comment (MasterSkepticista) [inline openfl/interface/plan.py:150]: As I mentioned in a previous commit, we want to avoid too much nesting. Please think of a broader approach that does not create too many edge cases. The goal of adding a `C901` check is to guard the code from getting too complex. By overriding C901 and adding nested conditionals only makes it exceptionally harder to test and validate corner cases

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]

File: openfl/component/aggregator/aggregator.py [hunk 13]
```diff
@@ -132,18 +134,59 @@ def __init__(
...
+
+            # TODO: Remove. Used in deprecated interactive and native APIs
+            self.best_tensor_dict: dict = {}
+            self.last_tensor_dict: dict = {}
+
+            self.collaborator_tensor_results = {}  # {TensorKey: nparray}}
+            if self._secure_aggregation_enabled:
+                from openfl.utilities.secagg.bootstrap import SecAggSetup
+
+                self.secagg = SecAggSetup(self.uuid, self.authorized_cols, self.tensor_db)
```
Comment (theakshaypant) [inline openfl/component/aggregator/aggregator.py:182]: These not be kept in a conditional since a default value is provided, seems like an unnecessary diff.

File: openfl/databases/tensor_db.py (hunk 9 above)
Comment (theakshaypant) [reply openfl/databases/tensor_db.py:268]: Perhaps consider using `LocalTensor` where weights can be disregarded by the `AggregationFunction`.

File: openfl/utilities/types.py [hunk 14]
```diff
@@ -12,6 +12,7 @@
 
 Metric = namedtuple("Metric", ["name", "value"])
 LocalTensor = namedtuple("LocalTensor", ["col_name", "tensor", "weight"])
+FALocalTensor = namedtuple("LocalTensor", ["col_name", "tensor"])
```
Comment (theakshaypant) [inline openfl/utilities/types.py:15]: Same comment as above as per the need for this new type.

File: openfl/interface/plan.py [hunk 15]
```diff
@@ -233,9 +234,10 @@ def freeze_plan(plan_config):
     plan = Plan()
     plan.config = Plan.parse(Path(plan_config), resolve=False).config
 
-    init_state_path = plan.config["aggregator"]["settings"]["init_state_path"]
+    init_state_path = plan.config["aggregator"]["settings"].get("init_state_path")
+    save_path = plan.config["aggregator"]["settings"].get("save_path")
 
-    if not Path(init_state_path).exists():
+    if any(path is not None and not Path(path).exists() for path in [init_state_path, save_path]):
```
Comment (theakshaypant) [inline openfl/interface/plan.py:240]: Seems simpler to have `if not (Path(init_state_path).exists() or Path(save_path).exists())` or do we foresee more paths being added in the future.

File: openfl/databases/tensor_db.py [hunk 16]
```diff
@@ -188,7 +188,9 @@ def get_aggregated_tensor(
         self,
         tensor_key: TensorKey,
         collaborator_weight_dict: dict,
+        collaborator_names: list,
         aggregation_function: AggregationFunction,
+        mode: str = "learning",
```
Comment (theakshaypant) [inline openfl/databases/tensor_db.py:193]: Similar to an earlier comment, db need (should) not be aware of the mode.

File: openfl/component/collaborator/collaborator.py [hunk 17]
```diff
@@ -204,59 +207,71 @@ def do_task(self, task, round_number) -> dict:
...
-        self.tensor_db.cache_tensor(local_output_tensor_dict)
+            global_output_tensor_dict = {}
+            tags = ("analysis",)
+            for key, metric in metrics.items():
+                global_output_tensor_dict[
+                    TensorKey(key, self.collaborator_name, round_number, False, tags)
+                ] = metric
 
+            self.tensor_db.cache_tensor(global_output_tensor_dict)
+            metrics = self.send_task_results(global_output_tensor_dict, round_number, task_name)
```
Comment (theakshaypant) [inline openfl/component/collaborator/collaborator.py:274]: Do not see the need for this conditional as we can pass `col_name`, `round_num` and `input_tensor_dict` to the task runner method where it can choose to not perform any other operation on these vars. 
As for `self.task_runner.get_required_tensorkeys_for_function`, it can just return an empty list for the analysis tasks.

**Review by theakshaypant — COMMENTED** [review]
Adding some comments from my first walkthrough. 
Still need to go through the workspace.

**Comment by tanwarsh** [comment]
Closing this as we went ahead without changing core component level changes.
